src/schemas/task.json

```diff
@@ -120,6 +120,9 @@
     },
     "Parameters": {
       "$ref": "paths.json#/definitions/asl_payload_template"
+    },
+    "Credentials": {
+      "$ref": "paths.json#/definitions/asl_payload_template"
```

Comment:
@massfords what do you think about using `asl_payload_template` for the `Credentials` field?

Reply:
I missed the addition of this field in the November 2022 update.

The issue I see is that the specification leaves it up to the interpreter to define the contents of the field while the examples from AWS only show a single field:

```
"Credentials": {
        "RoleArn": "arn:aws:iam::111122223333:role/LambdaRole"
      },
```

and another example using a path expression:

```
"Credentials": {
        "RoleArn.$": "$.roleArn"
      },
```

Maybe there should be a mode in the validator to not constrain the value vs a simple definition for what AWS commercial accepts. I'm also wondering if AWS accepts intrinsic functions for this field.

Reply:
I've added the tests, but while writing I think I found a bug there are 2 definitions for a Payload template.

`_asl_payload_template` and `asl_payload_template`. The one that I've used for `Credentials` and is being used by the `Parameters` is `asl_payload_template` in it's definition it allows types such as `number`, `string` and `bool`.
Meaning I can set the value for `Parameters` as a string. For example:
`{
  "Comment": "https://states-language.net/spec.html#parameters",
  "StartAt": "X",
  "States": {
    "X": {
      "Type": "Task",
      "Resource": "arn:aws:swf:us-east-1:123456789012:task:X",
      "Parameters": "This is an example"
      "End": true
    }
  }
}
`
When validating the state machine this will return `true`. Not sure if this is the expected behavior.
I've switched to using the `_asl_payload_template` definition for `Credentials` left the others unchanged.

Reply:
I don't see the switch to `_asl_payload_template` in the PR.

But actually it seems that `_asl_payload_template` is either an `_payload_template_object` or an array of `_asl_payload_template`.
I don't think we want to allow arrays for credentials, so I would go directly with `_payload_template_object` to only allow objects.

Reply:
About the `Parameters` field, I don't see any reaso to allow `number`, `string`, `bool` or an array.
I'll let @massfords decide wether to stick with the current `asl_payload_template` or restricting to objects with `_payload_template_object`

Reply:
From https://states-language.net:

> The value of "Parameters" MUST be a Payload Template which is a JSON object, whose input is the result of applying the InputPath to the raw input. If the "Parameters" field is provided, its payload, after the extraction and embedding, becomes the effective input.

Reply:
The difficulty here is that there isn't a formal schema for the shapes or grammar for the expression language. We're piecing these together from the spec and what actually works in AWS.

For example, the spec reference above says that the `Parameters` field is a JSON object but it should probably say it's a JSON value. It's valid in AWS Step Functions to have a string, number, boolean, or array of such as values. There were previous tickets about this since I had incorrectly assumed it had to be an `object`. See #101 

I'm not sure if the spec text should read `JSON value` as opposed to `JSON object`. These are better questions for the spec writers.

Here's a simple example that I just created and ran on AWS. It demonstrates that JSON values like `boolean`, `string`, `number`, and `array` work for `Parameters` alongside the more common `object`

```
{
  "StartAt": "Param with Object",
  "States": {
    "Param with Object": {
      "Type": "Pass",
      "Parameters": {
        "number": 42,
        "boolean": true,
        "string": "hello",
        "array": [
          42,
          true,
          "hello"
        ]
      },
      "Next": "Param with Path Refs",
      "ResultPath": "$.out"
    },
    "Param with Path Refs": {
      "Type": "Pass",
      "Parameters": {
        "number.$": "$.out.number",
        "boolean.$": "$.out.boolean",
        "string.$": "$.out.string"
      },
      "Next": "Param with boolean",
      "ResultPath": "$.out"
    },
    "Param with boolean": {
      "Type": "Pass",
      "Parameters": true,
      "Next": "Param with number"
    },
    "Param with number": {
      "Type": "Pass",
      "Parameters": 42,
      "Next": "Param with string"
    },
    "Param with string": {
      "Type": "Pass",
      "Parameters": "hello",
      "Next": "Param with array of bool"
    },
    "Param with array of bool": {
      "Type": "Pass",
      "Parameters": [
        true,
        false
      ],
      "Next": "Param with array of number"
    },
    "Param with array of number": {
      "Type": "Pass",
      "Parameters": [
        1,
        2
      ],
      "Next": "Param with array of strings"
    },
    "Param with array of strings": {
      "Type": "Pass",
      "Parameters": [
        "hello",
        "world"
      ],
      "Next": "Param with array of mixed"
    },
    "Param with array of mixed": {
      "Type": "Pass",
      "Parameters": [
        true,
        1,
        "world"
      ],
      "End": true
    }
  }
}
```

I believe I used an underscore as a prefix to denote `internal` types that shouldn't be referenced outside of the `paths.json` file. I think they're used for the recursive part of the definitions. 

- `asl_payload_template`: referenced by types for their `Parameters` field
- `_payload_template_object`: used recursively to define a JSON Object. This object has keys that end in `.$` and are valid path expressions OR keys that don't end in `.$` and are regular JSON values. 
- `_asl_payload_template`: **not referenced anywhere and should have been deleted as part of** #102 

I'm not aware of any other viable interpreters for step functions besides AWS. I think the few that existed were for offline testing and stopped being maintained once AWS released their local Docker runtime a few years ago. 

Until there are multiple interpreters, I'd prefer to see the validator focus on the AWS runtime as opposed to what it says in the language spec. The biggest difference here is the language spec saying the `Credentials` value is unconstrained but AWS requiring it to be a single field. 

The simplest approach to get this working is to have the validator only support what AWS supports and not leave the `Credentials` value unconstrained. From the AWS examples, this field MUST be an object with a single key `RoleArn` or `RoleArn.$`. If the latter, the value MUST be a string using the custom format `asl_ref_path`.

I think it's a reasonable assumption that a `RoleArn.$` value is a `Reference Path` as opposed to a `JSON Path` but I don't see where this is explicitly stated in the states language or AWS docs.

Reply:
In this case I'll just leave the `Credentials` field as `asl_payload_template` since that is a superset that covers both the AWS case with object with a key `RoleArn` or `RoleArn.$` and is compliant to the spec. In the future this could be extended to check AWS interpreter definition when a flag is set or the broad spec definition.

Reply:
Thanks for the explanation! I Think it's fair to use `asl_payload_template`